heliopy/data/mms.py

```diff
@@ -7,131 +7,1781 @@
 dl_url = mms_url + '/files/api/v1/download/science'
 
 
-def _validate_instrument(instrument):
-    allowed_instruments = ['afg', 'aspoc', 'dfg', 'dsp', 'edi',
-                           'edp', 'fgm', 'fpi', 'fields', 'scm', 'sdp', ]
-    if instrument not in allowed_instruments:
-        raise ValueError(
-            'Instrument {} not in list of allowed instruments: {}'.format(
-                instrument, allowed_instruments))
+class MMSDownloader(util.Downloader):
+    """
+    Download data from the Magnetospheric Multiscale (MMS) Science
+    Data Center (SDC). Data can be found at
+
+    https://lasp.colorado.edu/mms/sdc/
+
+    The following methods must be implemented by sub-classes:
+
+    - :meth:`Downloader.intervals()`: given a time interval, this
+      method should split the interval up into sub-intervals. Each of these
+      sub-intervals corresponds directly to a single file to download, store,
+      and read in.
+    - :meth:`Downloader.local_dir()`: given an interval, returns the
+      local directory in which the file is stored.
+    - :meth:`Downloader.fname()`: given an interval, returns the
+      local filename in which the file is stored.
+    - :meth:`Downloader.download()`: given an interval, download the data for
+      that interval.
+    - :meth:`Downloader.load_local_file()`: given an interval, load the local
+      file and return a :class:`pandas.DataFrame` object containing the data.
+    """
+    def __init__(self, sc=None, instr=None, mode=None, level=None,
+                 anc_product=None,
+                 data_type='science',
+                 data_root=None,
+                 dropbox_root=None,
+                 end_date=None,
+                 files=None,
+                 mirror_root=None,
+                 offline=False,
+                 optdesc=None,
+                 site='public',
+                 start_date=None,
+                 version=None):
+        """
+        Initialization method
+
+        Parameters
+        ----------
+        sc : str, list
+            Spacecraft ID.
+        instr : str, list
+            Instrument ID.
+        mode : str, list
+            Data rate mode.
+        level : str, list
+            Data product quality level. Setting level to None, "l2", or "l3"
+            automatically sets site to "public".
+        anc_product : bool
+            Set to True if you want ancillary data. Automatically sets
+            "data_type" to "ancillary".
+        data_type : str
+            Type of information requested from the SDC:
+                "science" - Science data
+                "hk" - Housekeeping data
+                "ancillary" - attitude and emphemeris data
+                "abs_selections" - Automated burst segment selections
+                "sitl_selections" - Burst selections made by the SITL
+                "gls_selections_*" - Burst selections made by the ground loop
+        data_root : str
```

Comment:
I haven't added this option to other downloaders, so I think it would be best not to add it here. A user can always change `heliopy.config['download_dir']` to change where data downloads to.